File: akka-docs/src/test/scala/docs/stream/operators/sourceorflow/MapAsyncs.scala
/*
 * Copyright (C) 2020-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.stream.operators.sourceorflow

import akka.Done
import akka.NotUsed
import akka.actor.ActorSystem
import akka.stream.scaladsl.Flow
import akka.stream.scaladsl.Keep
import akka.stream.scaladsl.Sink
import akka.stream.scaladsl.Source
import akka.util.Timeout

import scala.collection.immutable
import scala.concurrent.ExecutionContextExecutor
import scala.concurrent.Future
import scala.concurrent.duration._
import scala.util.Random

object CommonMapAsync {
  sealed trait Event {
    def sequenceNumber: Int
  }

  object Event {
    def apply(sequenceNumber: Int): PlainEvent = PlainEvent(sequenceNumber)
  }

  case class PlainEvent(sequenceNumber: Int) extends Event
  case class EntityEvent(entityId: Int, sequenceNumber: Int) extends Event

  // dummy objects for our pretend Kafka
  val settings = new AnyRef
  val subscription = new AnyRef

  // for access from java
  def consumer = Consumer
  def committer = Committer

  object Consumer {
    // almost but not quite like Alpakka Kafka
    def committableSource(settings: AnyRef, subscription: AnyRef): Source[EntityEvent, NotUsed] =
      // pro forma to use the dummy arguments
      if (settings != null && subscription != null) {
        Source
          .fromIterator { () =>
            Iterator.from(1)
          }
          .statefulMap(() => immutable.Map.empty[Int, Int])(
            { (countsPerEntity, n) =>
              val entityId =
                if (Random.nextBoolean() || countsPerEntity.isEmpty) Random.nextInt(n)
                else {
                  val m = Random.nextInt(countsPerEntity.size) + 1
                  countsPerEntity.keys.iterator.take(m).drop(m - 1).next()
                }
              val seqNr = countsPerEntity.getOrElse(entityId, 0) + 1
              (countsPerEntity + (entityId -> seqNr), EntityEvent(entityId, seqNr))
            }, { _ =>
              None
            })
      } else throw new AssertionError("pro forma")

    // likewise ape the Alpakka Kafka API
    def plainSource(settings: AnyRef, subscription: AnyRef): Source[PlainEvent, NotUsed] =
      // pro forma use the dummy arguments
      if (settings != null && subscription != null) {
        Source
          .fromIterator { () =>
            Iterator.from(1)
          }
          .map(Event.apply _)
      } else throw new AssertionError("pro forma")
  }

  object Committer {
    def sink(prependTo: String): Sink[String, Future[Done]] =
      Flow[String].map(in => s"$prependTo emitted $in").toMat(Sink.foreach(println _))(Keep.right)
  }

  implicit val sys: ActorSystem = ActorSystem("mapAsync-stream")
  implicit val exCtx: ExecutionContextExecutor = sys.dispatcher
  implicit val timeout: Timeout = 3.seconds

  // #mapasync-strict-order
  // #mapasync-concurrent
  // #mapasyncunordered

  val events: Source[Event, NotUsed] = Consumer.plainSource(settings, subscription).throttle(1, 50.millis)

  def eventHandler(event: Event): Future[Int] = {
    println(s"Processing event $event...")
    // #mapasync-strict-order
    // #mapasync-concurrent
    // #mapasyncunordered
    val result =
      if (Random.nextInt(5) == 0) {
        akka.pattern.after(500.millis)(Future.successful(event.sequenceNumber))
      } else {
        Future.successful(event.sequenceNumber)
      }
    result.map { x =>
      println(s"Completed processing $x")
      x
    }
    // #mapasync-strict-order
    // #mapasync-concurrent
    // #mapasyncunordered
  }
  // #mapasync-strict-order
  // #mapasync-concurrent
  // #mapasyncunordered

}

object MapAsyncStrictOrder extends App {
  import CommonMapAsync._
  // #mapasync-strict-order

  events
    .mapAsync(1) { in =>
      eventHandler(in)
    }
    .map { in =>
      println(s"`mapAsync` emitted event number: $in")
    }
    // #mapasync-strict-order
    .runWith(Sink.ignore)

}

object MapAsync extends App {
  import CommonMapAsync._
  // #mapasync-concurrent

  events
    .mapAsync(3) { in =>
      eventHandler(in)
    }
    .map { in =>
      println(s"`mapAsync` emitted event number: $in")
    }
    // #mapasync-concurrent
    .runWith(Sink.ignore)

}

object MapAsyncUnordered extends App {
  import CommonMapAsync._
  // #mapasyncunordered

  events
    .mapAsyncUnordered(3) { in =>
      eventHandler(in)
    }
    .map { in =>
      println(s"`mapAsyncUnordered` emitted event number: $in")
    }
    // #mapasyncunordered
    .runWith(Sink.ignore)

}

object MapAsyncPartitioned extends App {
  import CommonMapAsync._

  // pretending this is an Alpakka Kafka-style CommitterSettings
  val commitSettings = "`mapAsyncPartitioned`"

  // #mapAsyncPartitioned
  val eventsForEntities: Source[EntityEvent, NotUsed] =
    Consumer.committableSource(settings, subscription).take(1000)

  val partitioner: EntityEvent => Int = { event =>
    val partition = event.entityId
    println(s"Assigned event $event to partition $partition")
    partition
  }

  eventsForEntities.zipWithIndex
    .map {
      case (event, count) =>
        println(s"Received event $event at offset $count from message broker")
        event
    }
    .mapAsyncPartitioned(10, 1)(partitioner) { (event, partition) =>
      println(s"Processing event $event from partition $partition")

      // processing result is "partition-sequenceNr"
      // val processEvent: (EntityEvent, Int) => Future[String]
      processEvent(event, partition).map { x =>
        println(s"Completed processing $x")
        x
      }
    }
    // for the purpose of this example, will print every element, prepended with "`mapAsyncPartitioned` emitted "
    .runWith(Committer.sink(commitSettings))
    // #mapAsyncPartitioned
    .flatMap { _ =>
      // give the printlns a chance to print
      akka.pattern.after(1.second)(Future.unit)
    }
    .foreach { _ =>
      System.out.flush()
      sys.terminate()
    }

  def processEvent(event: EntityEvent, partition: Int): Future[String] = {
    val fut = Future.successful(s"$partition-${event.sequenceNumber}")
    if (Random.nextBoolean()) {
      akka.pattern.after(Random.nextInt(1000).millis)(fut)
    } else fut
  }
}

File: akka-docs/src/test/java/jdocs/stream/operators/sourceorflow/MapAsyncs.java
/*
 * Copyright (C) 2020-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.stream.operators.sourceorflow;

import akka.Done;
import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.japi.Pair;
import akka.japi.function.Function;
import akka.pattern.Patterns;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import docs.stream.operators.sourceorflow.CommonMapAsync;
import java.time.Duration;
import java.util.Optional;
import java.util.Random;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import scala.jdk.javaapi.FutureConverters;

public class MapAsyncs {

  private static final Random random = new Random();

  // dummy objects for our pretend Kafka...
  private Object settings = new Object();
  private Object subscription = new Object();

  // #mapasync-strict-order
  // #mapasync-concurrent
  // #mapasyncunordered

  private final Source<Event, NotUsed> events =
      Consumer.plainSource(settings, subscription).throttle(1, Duration.ofMillis(50));
  // #mapasync-strict-order
  // #mapasync-concurrent
  // #mapasyncunordered

  private final ActorSystem system = ActorSystem.create("mapAsync-operator-examples");

  public MapAsyncs() {}

  // #mapasync-strict-order
  // #mapasync-concurrent
  // #mapasyncunordered

  public CompletionStage<Integer> eventHandler(Event in) throws InterruptedException {
    System.out.println("Processing event number " + in + "...");
    // ...
    // #mapasync-strict-order
    // #mapasync-concurrent
    // #mapasyncunordered
    CompletionStage<Integer> cs;
    if (random.nextInt(5) == 0) {
      cs =
          Patterns.after(
              Duration.ofMillis(500),
              system,
              () -> CompletableFuture.completedFuture(in.sequenceNumber()));
    } else {
      cs = CompletableFuture.completedFuture(in.sequenceNumber());
    }
    return cs.thenApply(
        i -> {
          System.out.println("Completed processing " + i.intValue());
          return i;
        });
    // #mapasync-strict-order
    // #mapasync-concurrent
    // #mapasyncunordered
  }
  // #mapasync-strict-order
  // #mapasync-concurrent
  // #mapasyncunordered

  private void runStrictOrder() {
    // #mapasync-strict-order

    events
        .mapAsync(1, this::eventHandler)
        .map(in -> "`mapSync` emitted event number " + in.intValue())
        .runWith(Sink.foreach(str -> System.out.println(str)), system);
    // #mapasync-strict-order
  }

  private void run() {
    // #mapasync-concurrent

    events
        .mapAsync(10, this::eventHandler)
        .map(in -> "`mapSync` emitted event number " + in.intValue())
        .runWith(Sink.foreach(str -> System.out.println(str)), system);
    // #mapasync-concurrent
  }

  private void runUnordered() {
    // #mapasyncunordered

    events
        .mapAsyncUnordered(10, this::eventHandler)
        .map(in -> "`mapSync` emitted event number " + in.intValue())
        .runWith(Sink.foreach(str -> System.out.println(str)), system);
    // #mapasyncunordered
  }

  private void runPartitioned() {
    // pretending this is an Alpakka Kafka-style CommitterSettings
    String commitSettings = "`mapAsyncPartitioned`";

    // #mapAsyncPartitioned
    Function<EntityEvent, Integer> partitioner =
        (event) -> {
          Integer partition = Integer.valueOf(event.entityId);
          System.out.println("Assigned event " + event + " to partition " + event.entityId);
          return partition;
        };

    Consumer.committableSource(settings, subscription)
        .take(1000)
        .statefulMap(
            () -> Integer.valueOf(0),
            (count, event) -> {
              System.out.println(
                  "Received event " + event + " at offset " + count + " from message broker");
              return Pair.create(count + 1, event);
            },
            (count) -> Optional.empty())
        .mapAsyncPartitioned(
            10, // parallelism
            1, // perPartition
            partitioner,
            (event, partition) -> {
              System.out.println("Processing event " + event + " from partition " + partition);

              // processing result is "partition-sequenceNr"
              // public CompletionStage<String> processEvent(EntityEvent event, Integer partition)
              CompletionStage<String> cs = processEvent(event, partition);

              return cs.thenApply(
                  s -> {
                    System.out.println("Completed processing " + s);
                    return s;
                  });
            })
        // for the purpose of this example, will print every element, prepended with
        // "`mapAsyncPartitioned` emitted "
        .runWith(Committer.sink(commitSettings), system);
    // #mapAsyncPartitioned
  }

  public static void main(String[] args) {
    MapAsyncs mapAsyncs = new MapAsyncs();
    if (args[0].equals("partitioned")) {
      mapAsyncs.runPartitioned();
    } else if (args[0].equals("strict")) {
      mapAsyncs.runStrictOrder();
    } else if (args[0].equals("unordered")) {
      mapAsyncs.runUnordered();
    } else {
      mapAsyncs.run();
    }
  }

  static interface Event {
    public int sequenceNumber();
  }

  static class PlainEvent implements Event {
    public final int _sequenceNumber;

    public PlainEvent(int sequenceNumber) {
      this._sequenceNumber = sequenceNumber;
    }

    public int sequenceNumber() {
      return _sequenceNumber;
    }

    @Override
    public String toString() {
      return "Event(" + _sequenceNumber + ')';
    }
  }

  static class EntityEvent implements Event {
    public final int entityId;
    public final int _sequenceNumber;

    public EntityEvent(int entityId, int sequenceNumber) {
      this.entityId = entityId;
      this._sequenceNumber = sequenceNumber;
    }

    public int sequenceNumber() {
      return _sequenceNumber;
    }

    @Override
    public String toString() {
      return "EntityEvent(" + entityId + ", " + _sequenceNumber + ")";
    }
  }

  // Pretend to be asking an entity from an event... response is sometimes delayed by up to a second
  public CompletionStage<String> processEvent(EntityEvent event, Integer partition) {
    CompletableFuture<String> cf =
        CompletableFuture.completedFuture(partition.toString() + "-" + event.sequenceNumber());

    if (random.nextBoolean()) {
      return Patterns.after(Duration.ofMillis(random.nextInt(1000)), system, () -> cf);
    } else {
      return cf;
    }
  }

  private static class Consumer {
    // almost but not quite, like Alpakka Kafka...
    public static Source<EntityEvent, NotUsed> committableSource(
        Object settings, Object subscription) {
      return CommonMapAsync.consumer()
          .committableSource(settings, subscription)
          .asJava()
          .map(scalaEvent -> new EntityEvent(scalaEvent.entityId(), scalaEvent.sequenceNumber()));
    }

    // likewise ape the Alpakka Kafka API
    public static Source<Event, NotUsed> plainSource(Object settings, Object subscription) {
      return CommonMapAsync.consumer()
          .plainSource(settings, subscription)
          .asJava()
          .map(scalaEvent -> new PlainEvent(scalaEvent.sequenceNumber()));
    }
  }

  private static class Committer {
    public static Sink<String, CompletionStage<Done>> sink(String prependTo) {
      return CommonMapAsync.committer()
          .sink(prependTo)
          .asJava()
          .mapMaterializedValue(FutureConverters::asJava);
    }
  }
}
